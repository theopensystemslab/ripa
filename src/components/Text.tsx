import { Button } from "@material-ui/core";
import Box from "@material-ui/core/Box";
import Input from "@material-ui/core/Input";
import InputLabel from "@material-ui/core/InputLabel";
import { useFormik } from "formik";
import * as React from "react";
import FocusWithin from "react-focus-within";

import Messages from "../shared/components/submit-messages";
import Question from "./Question";

interface IMinMax {
  min?: number;
  max?: number;
}
interface IText {
  title: string;
  type: string;
  name: string;
  multiline?: boolean;
  label?: string;
  fullWidth?: boolean;
  required?: boolean;
  placeholder?: string;
  unit?: string;
  min?: string;
  max?: string;
  maxWords?: number;
  inputProps?: IMinMax;
  includeSubmit?: boolean;
}

export const Text: React.FC<IText> = ({
  title,
  label = false,
  fullWidth = true,
  placeholder = "",
  multiline = false,
  required = false,
  name = "",
  type = "",
  unit = "",
  maxWords,
  inputProps = {
    min: 0,
    max: Infinity
  },
  includeSubmit = false
}) => {
  const [errorMessageVisible, setErrorMessageVisible] = React.useState(false);
  const [successMessageVisible, setSuccessMessageVisible] = React.useState(
    false
  );
  const [count, setCount] = React.useState(0);
  const diff = maxWords - count;
  const formik = useFormik({
    initialValues: {
      [name]: ""
    },
    onSubmit: values => {
      if (diff > 0) {
        console.log(JSON.stringify(values, null, 2));
        setSuccessMessageVisible(true);
        setErrorMessageVisible(false);
      } else {
        setSuccessMessageVisible(false);
        setErrorMessageVisible(true);
      }
    }
  });

  return (
    <FocusWithin>
      {({ isFocused, getFocusProps }) => (
        <div {...getFocusProps()}>
          <Box py={4} maxWidth={480}>
            <form data-testid="textForm" onSubmit={formik.handleSubmit}>
              <Box mb={1.5}>
                <Question inFocus={isFocused}>{title}</Question>
              </Box>
              <Box>
                {label && <InputLabel htmlFor={name}>{label}</InputLabel>}
                <Input
                  placeholder={placeholder}
                  fullWidth={fullWidth}
                  multiline={multiline}
                  name={name}
                  id={name}
                  type={type}
                  onChange={e => {
                    if (maxWords) {
                      setCount(e.target.value.split(" ").length - 1);
                    }
                    formik.handleChange(e);
                  }}
                  rows={multiline ? 5 : 1}
                  value={formik.values[name]}
                  required={required}
                  {...inputProps}
                />
                {maxWords && (
                  <Box
                    fontSize="caption.fontSize"
                    color="text.secondary"
                    pt={1}
                    textAlign="right"
                  >
                    <span>
                      {" "}
                      {diff >= 0
                        ? `${diff} words Remaining`
                        : `0 words Remaining`}{" "}
                    </span>
                  </Box>
                )}
                {unit && (
                  <Box
                    component="span"
                    fontSize="caption.fontSize"
                    style={{ verticalAlign: "bottom", lineHeight: "40px" }}
                    pl={1.5}
                  >
                    {unit}
                  </Box>
                )}
                <div>
                  {errorMessageVisible && formik.touched ? (
                    <Messages
                      type="error"
                      message="You exceeded the max number of words allowed!"
                    />
                  ) : null}
                </div>
              </Box>
              {includeSubmit && (
                <Button variant="contained" color="primary" type="submit">
                  Save and Continue
                </Button>
              )}
              <div>
                {successMessageVisible ? (
                  <Messages
                    type="success"
                    message="Form submitted successfully"
                  />
                ) : null}
              </div>
            </form>
          </Box>
        </div>
      )}
    </FocusWithin>
  );
};

export default Text;
