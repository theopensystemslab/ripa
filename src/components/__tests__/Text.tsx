import { fireEvent, render } from "@testing-library/react";
import * as React from "react";

import Text from "../Text";

describe("Text Component", () => {
  let title: string;
  let name: string;
  let type: string;
  let placeholder: string;
  let multiline: boolean;
  let maxWords: number;
  let label: string;
  // let unit: string;

  beforeEach(() => {
    title = "test title";
    name = "test name";
  });
  describe("of type short text", () => {
    beforeEach(() => {
      placeholder = "short text input placeholder";
      type = "text";
      label = "short text input label";
    });
    it("should render snapshot", () => {
      const { asFragment } = render(
        <Text
          title={title}
          label={label}
          placeholder={placeholder}
          name={name}
          type={type}
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });
    it("should have the right label and placeholder text", async () => {
      const { findByPlaceholderText, findByLabelText } = render(
        <Text
          title={title}
          label={label}
          placeholder={placeholder}
          name={name}
          type={type}
        />
      );
      expect(await findByPlaceholderText(placeholder)).toBeInTheDocument();
      expect(await findByLabelText(label)).toBeInTheDocument();
    });
    it("should have a value if an text is entered", async () => {
      const value = "Lorem ipsum dolor sit amet, consectetur adipiscing";
      const { findByPlaceholderText, getByPlaceholderText } = render(
        <Text
          title={title}
          label={label}
          placeholder={placeholder}
          name={name}
          type={type}
        />
      );
      fireEvent.change(getByPlaceholderText(placeholder), {
        target: { value }
      });
      expect(await findByPlaceholderText(placeholder)).toHaveValue(value);
    });
  });
  describe("Of type long text", () => {
    beforeEach(() => {
      type = "text";
      maxWords = 10;
      placeholder = "long text input placeholder";
      multiline = true;
      label = "long text input label";
    });
    it("should render snapshot", () => {
      const { asFragment } = render(
        <Text
          title={title}
          label={label}
          placeholder={placeholder}
          multiline={multiline}
          maxWords={maxWords}
          name={name}
          type={type}
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });
    it("should have the right label and placeholder text", async () => {
      const { findByPlaceholderText, findByLabelText } = render(
        <Text
          title={title}
          label={label}
          placeholder={placeholder}
          multiline={multiline}
          maxWords={maxWords}
          name={name}
          type={type}
        />
      );
      expect(await findByPlaceholderText(placeholder)).toBeInTheDocument();
      expect(await findByLabelText(label)).toBeInTheDocument();
    });
    it("should have a value if an text is entered", async () => {
      const value =
        "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.";
      const { findByPlaceholderText, getByPlaceholderText } = render(
        <Text
          title={title}
          label={label}
          placeholder={placeholder}
          multiline={multiline}
          maxWords={maxWords}
          name={name}
          type={type}
        />
      );
      fireEvent.change(getByPlaceholderText(placeholder), {
        target: { value }
      });
      expect(await findByPlaceholderText(placeholder)).toHaveValue(value);
    });
    it("should render number of words remaining", async () => {
      const value =
        "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.";
      const { findByText, getByPlaceholderText } = render(
        <Text
          title={title}
          label={label}
          placeholder={placeholder}
          multiline={multiline}
          maxWords={maxWords}
          name={name}
          type={type}
        />
      );
      fireEvent.change(getByPlaceholderText(placeholder), {
        target: { value }
      });
      expect(await findByText(/0 words Remaining/i)).toBeInTheDocument();
    });
    it("should render FAIL message when submit is enabled and no of words is OUT OF range", async () => {
      const value =
        "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.";
      const { findByText, getByTestId, getByPlaceholderText } = render(
        <Text
          title={title}
          label={label}
          placeholder={placeholder}
          multiline={multiline}
          maxWords={maxWords}
          name={name}
          type={type}
          includeSubmit={true}
        />
      );
      fireEvent.change(getByPlaceholderText(placeholder), {
        target: { value }
      });
      fireEvent.submit(getByTestId("textForm"));

      expect(
        await findByText(/you exceeded the max number of words allowed!/i)
      ).toBeInTheDocument();
    });
    it("should render success message when submit is enabled and no of words is within range", async () => {
      const value = "Lorem ipsum dolor sit amet";
      const { findByText, getByTestId, getByPlaceholderText } = render(
        <Text
          title={title}
          label={label}
          placeholder={placeholder}
          multiline={multiline}
          maxWords={maxWords}
          name={name}
          type={type}
          includeSubmit={true}
        />
      );
      fireEvent.change(getByPlaceholderText(placeholder), {
        target: { value }
      });
      fireEvent.submit(getByTestId("textForm"));
      expect(
        await findByText(/form submitted successfully/i)
      ).toBeInTheDocument();
    });
  });
  describe("of type Email", () => {
    beforeEach(() => {
      type = "email";
      label = "Email text input label";
      placeholder = "Email text input placeholder";
    });
    it("should render snapshot", () => {
      const { asFragment } = render(
        <Text
          title={title}
          placeholder={placeholder}
          name={name}
          type={type}
          label={label}
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });
    it("should have the right label and placeholder text", async () => {
      const { findByPlaceholderText, findByLabelText } = render(
        <Text
          title={title}
          placeholder={placeholder}
          name={name}
          type={type}
          label={label}
        />
      );
      expect(await findByPlaceholderText(placeholder)).toBeInTheDocument();
      expect(await findByLabelText(label)).toBeInTheDocument();
    });
    it("should have a value if an email is entered", async () => {
      const { findByPlaceholderText, getByPlaceholderText } = render(
        <Text
          title={title}
          placeholder={placeholder}
          name={name}
          type={type}
          label={label}
        />
      );
      fireEvent.change(getByPlaceholderText(placeholder), {
        target: { value: "test@test.com" }
      });
      expect(await findByPlaceholderText(placeholder)).toHaveValue(
        "test@test.com"
      );
    });
  });
  describe("of type Number", () => {
    beforeEach(() => {
      type = "number";
      label = "number input label";
      placeholder = "number input placeholder";
    });
    it("should render snapshot", () => {
      const { asFragment } = render(
        <Text
          title={title}
          placeholder={placeholder}
          name={name}
          type={type}
          label={label}
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });
    it("should have the right label and placeholder text", async () => {
      const { findByPlaceholderText, findByLabelText } = render(
        <Text
          title={title}
          placeholder={placeholder}
          name={name}
          type={type}
          label={label}
        />
      );
      expect(await findByPlaceholderText(placeholder)).toBeInTheDocument();
      expect(await findByLabelText(label)).toBeInTheDocument();
    });
    it("should render the unit when specified", async () => {
      const unit = "meters";
      const { findByText } = render(
        <Text
          title={title}
          placeholder={placeholder}
          name={name}
          type={type}
          unit={unit}
          label={label}
        />
      );
      expect(await findByText("meters")).toBeInTheDocument();
    });
    it("should not have any value if text is entered", async () => {
      const { findByPlaceholderText, getByPlaceholderText } = render(
        <Text
          title={title}
          placeholder={placeholder}
          name={name}
          type={type}
          label={label}
        />
      );
      fireEvent.change(getByPlaceholderText(placeholder), {
        target: { value: "Test text" }
      });
      expect(await findByPlaceholderText(placeholder)).toHaveValue(null);
    });
    it("should have a value if a number is entered", async () => {
      const { findByPlaceholderText, debug, getByPlaceholderText } = render(
        <Text
          title={title}
          placeholder={placeholder}
          name={name}
          type={type}
          label={label}
        />
      );
      fireEvent.change(getByPlaceholderText(placeholder), {
        target: { value: 123 }
      });
      expect(await findByPlaceholderText(placeholder)).toHaveValue(123);
    });
  });
});
