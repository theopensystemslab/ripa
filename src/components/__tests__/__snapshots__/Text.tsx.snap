// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Component Of type long text should render snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="MuiBox-root MuiBox-root-269"
    >
      <form
        data-testid="textForm"
      >
        <div
          class="MuiBox-root MuiBox-root-270"
        >
          <div
            class="makeStyles-root-271"
          >
            <div
              class="MuiTypography-root makeStyles-question-272 MuiTypography-h4"
            >
              <strong>
                test title
              </strong>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-304"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated"
            for="test name"
          >
            long text input label
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-multiline MuiInput-multiline"
            max="Infinity"
            min="0"
          >
            <textarea
              class="MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline MuiInput-inputMultiline"
              id="test name"
              name="test name"
              placeholder="long text input placeholder"
              rows="5"
            />
          </div>
          <div
            class="MuiBox-root MuiBox-root-357"
          >
            <span>
               10 words Remaining 
            </span>
          </div>
          <div />
        </div>
        <div />
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Text Component of type Email should render snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="MuiBox-root MuiBox-root-893"
    >
      <form
        data-testid="textForm"
      >
        <div
          class="MuiBox-root MuiBox-root-894"
        >
          <div
            class="makeStyles-root-895"
          >
            <div
              class="MuiTypography-root makeStyles-question-896 MuiTypography-h4"
            >
              <strong>
                test title
              </strong>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-928"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated"
            for="test name"
          >
            Email text input label
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth"
            max="Infinity"
            min="0"
          >
            <input
              class="MuiInputBase-input MuiInput-input"
              id="test name"
              name="test name"
              placeholder="Email text input placeholder"
              rows="1"
              type="email"
              value=""
            />
          </div>
          <div />
        </div>
        <div />
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Text Component of type Number should render snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="MuiBox-root MuiBox-root-1160"
    >
      <form
        data-testid="textForm"
      >
        <div
          class="MuiBox-root MuiBox-root-1161"
        >
          <div
            class="makeStyles-root-1162"
          >
            <div
              class="MuiTypography-root makeStyles-question-1163 MuiTypography-h4"
            >
              <strong>
                test title
              </strong>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-1195"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated"
            for="test name"
          >
            number input label
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth"
            max="Infinity"
            min="0"
          >
            <input
              class="MuiInputBase-input MuiInput-input"
              id="test name"
              name="test name"
              placeholder="number input placeholder"
              rows="1"
              type="number"
              value=""
            />
          </div>
          <div />
        </div>
        <div />
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Text Component of type short text should render snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="MuiBox-root MuiBox-root-2"
    >
      <form
        data-testid="textForm"
      >
        <div
          class="MuiBox-root MuiBox-root-3"
        >
          <div
            class="makeStyles-root-4"
          >
            <div
              class="MuiTypography-root makeStyles-question-5 MuiTypography-h4"
            >
              <strong>
                test title
              </strong>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-37"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated"
            for="test name"
          >
            short text input label
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth"
            max="Infinity"
            min="0"
          >
            <input
              class="MuiInputBase-input MuiInput-input"
              id="test name"
              name="test name"
              placeholder="short text input placeholder"
              rows="1"
              type="text"
              value=""
            />
          </div>
          <div />
        </div>
        <div />
      </form>
    </div>
  </div>
</DocumentFragment>
`;
